sdk/cosmos/azure_data_cosmos/src/routing/async_cache.rs

```diff
@@ -0,0 +1,282 @@
+    value: V,
+    expires_at: Instant,
+}
+
+impl<V> CacheEntry<V> {
+    fn new(value: V, ttl: Duration) -> Self {
+        Self {
+            value,
+            expires_at: Instant::now() + ttl,
+        }
+    }
+
+    fn is_expired(&self) -> bool {
+        Instant::now() >= self.expires_at
+    }
+}
+
+/// A generic async cache with TTL support.
+#[derive(Clone)]
+pub struct AsyncCache<K, V>
+where
+    K: Eq + Hash + Clone,
+    V: Clone,
+{
+    store: Arc<RwLock<HashMap<K, CacheEntry<V>>>>,
+    ttl: Duration,
+}
+
+impl<K, V> AsyncCache<K, V>
+where
+    K: Eq + Hash + Clone + Send + Sync + 'static,
+    V: Clone + Send + Sync + 'static,
+{
+    /// Creates a new `AsyncCache` with the specified TTL.
+    pub fn new(ttl: Duration) -> Self {
+        Self {
+            store: Arc::new(RwLock::new(HashMap::new())),
+            ttl,
+        }
+    }
+
+    /// Gets a value from the cache, or computes it using the provided async function if not present or expired
+    ///
+    /// When the entry is expired or a force_refresh is requested, the cache is automatically updated
+    /// with the newly computed value.
+    ///
+    /// # Arguments
+    /// * `key` - The cache key to look up
+    /// * `force_refresh` - If true, bypass cache and compute fresh value even if cached value is valid
+    /// * `compute` - Async function to compute the value if not cached or force refresh is requested
+    pub async fn get<F, Fut, E>(&self, key: K, force_refresh: bool, compute: F) -> Result<V, E>
+    where
+        F: FnOnce() -> Fut,
+        Fut: std::future::Future<Output = Result<V, E>>,
+    {
+        // Fast path: check if value exists and is not expired (read lock)
+        // Skip this if force_refresh is true
+        if !force_refresh {
+            let store = self.store.read().await;
+            if let Some(entry) = store.get(&key) {
+                if !entry.is_expired() {
+                    return Ok(entry.value.clone());
+                }
+            }
+        }
+
+        // Slow path: value missing, expired, or force refresh requested - need to compute (write lock)
+        let mut store = self.store.write().await;
+
+        // Double-check after acquiring write lock (another task might have updated it)
+        // Only skip recompute if not force_refresh and entry is still valid
+        if !force_refresh {
+            if let Some(entry) = store.get(&key) {
+                if entry.is_expired() {
+                    // Remove the entry from the cache.
+                    store.remove(&key);
+                } else {
+                    // Another task updated it while we waited for the lock
+                    return Ok(entry.value.clone());
+                }
+            }
+        } else {
+            // force_refresh is true, remove existing entry to ensure fresh computation
+            store.remove(&key);
+        }
+
+        // Compute new value
+        let value = compute().await?;
```

Comment:
I think we can use per-key locking pretty easily here, but I'll save that for a follow-up refactoring PR.